group/assignment6/RearrangingCars.java

```diff
@@ -0,0 +1,89 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Identifiers for parking spaces: A,B,C,...
+ * Identifiers for cars: 1,2,3,...
+ * <p>
+ * State of the parking lot is represented by a map, which has the spaces as keys and the current car on the space as values
+ * If there is no car the value of the space is 0
+ */
+public class RearrangingCars {
+
+    /**
+     * Gets to valid parking lot states and generates a sequence of moves from the start state to the end state
+     * (Valid = same number of spaces/cars and just one free space in each state)
+     *
+     * @param startState start state of the parking lot
+     * @param endState   end state of the parking lot
+     * @return sequence of moves from the start state to the end state
+     */
+    public static List<Step> rearrangeCars(Map<Character, Integer> startState, Map<Character, Integer> endState) {
+        List<Step> steps = new LinkedList<>();
+        while (!startState.equals(endState)) {
+            Step step = findNextStep(startState, endState);
+            if (step != null) {
+                steps.add(step);
+                startState.put(step.from, 0);
+                startState.put(step.to, step.car);
+            }
+        }
+        return steps;
+    }
+
+    /**
+     * Finds the best next step, so that one more position is correct
+     *
+     * @param start current state of the parking lot
+     * @param end   end state of the parking lot
+     * @return next step
+     */
+    private static Step findNextStep(Map<Character, Integer> start, Map<Character, Integer> end) {
+        // find free space in start state
+        char freeSpace = 'A';
```

Comment:
what if you do not have any free space. I think you will loose the car existing on space 'A'. Let me know if i am right.

Reply:
Yes, you are right. I assumed I'll only get valid parking states. So that they will all have an empty spot. Otherwise the tasked wouldn't be solvable, because we are not allowed to swap two cars. But I could add a check at the beginning to make sure the configurations are valid.